File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-2.java
class Solution {
    public int findMin(int[] nums) {
        int n = nums.length;
        int r = n-1;
        int l = 0;
        int res = 1001;
        while(l <= r) {
            int mid = l + (r - l) / 2;
            res = Math.min(res, nums[mid]);
            if(nums[mid] > nums[r]) {
                l = mid + 1;
            } else {
                r = mid - 1;
            }
        }
        return res;
    }
}

File: Data Structures & Algorithms/permutations/submission-2.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        res = []
        visited = [False] * len(nums)

        def dfs(perm):
            if len(perm) == len(nums):
                res.append(perm.copy())
                return

            for j in range(len(nums)):
                if not visited[j]:
                    perm.append(nums[j])
                    visited[j] = True
                    dfs(perm)
                    visited[j] = False
                    perm.pop()
        
        dfs([])
        return res
